### PR TITLE
gsdx: Generic config load

### PR DESCRIPTION
**Pull request by willkuer**
This commit enables the possibility to load settings in the same datatype as they are expected. This should increase flexability, readability (and possibly uncritical speed).

This is achieved by making GsdxApp::GetConfig generic. The two current implementations are given as specialization. Additionally as third specialization a boolean value can now be read.

The PR consists currently of three commits:
1. making GsdxApp:GetConfig generic as described above
2. Fixing warnings that appear if somebody requests bool = theApp.GetConfig( [ * ], 0/1) as an implicit conversion from int to bool happens.
3. Fixing if somebody requests bool = !!theApp.GetConfig( [ * ], 0/1) as one can now directly get the boolean value by bool = theApp.GetConfig( [*], false/true)

This pr is open for discussion. It is not in mergable state. I just wanted to know if there is general interest in implementing such a system. If you like it this PR is probably the first of three to make GSSettings and config load more flexible and readable. ( the other two are SetConfig and GSSetting uint-issue)

### CODE REVIEW

**Comment by gregory38** [comment]
Did you test the linux side? Honestly I think the bool stuff will work with any templates. Implicit casting ought to handle it.

**Comment by willkuer** [comment]
No linux compilation I haven't done. I guess there are still some warnings and possibly errors. That's why I said its not in mergable state. It's just that clean build on my VM takes ~1 hour I guess. Writing this took much less. I would do linux side if you like the general approach.

About bool one could probably as well static_cast to int and vice versa. I am not sure which solution looks better or is faster. In a [c++ shell](http://cpp.sh/) static_cast worked as expected and to my knowledge it should be well defined behavior. On the other hand ternary operator and comparison to non-zero is for sure well defined and easily readable.

Actually in my mind bool specialization and remplate should've called int specialization after 'casting' to int. I will change that. There should be only one call to GetPrivateProfileInt.

**Comment by willkuer** [comment]
Ok. That's the idea.

GSSettings and ConfigLoad and Save are now generic on the surface. This reduces casting in consumer code.

On the fly one warning was fixed where a signed integer was saved as GSSetting.id (uint). Now the behavior for unsigned ints is well defined. Additionally usage of (strongly typed) enums is much easier/looks better.

Code compiles in Linux/Windows.

File: plugins/GSdx/GSCaptureDlg.cpp [hunk 1]
```diff
@@ -64,7 +64,7 @@ void GSCaptureDlg::OnInit()
 
 	m_codecs.clear();
 
-	_bstr_t selected = theApp.GetConfig("CaptureVideoCodecDisplayName", "").c_str();
+	_bstr_t selected = CComBSTR(theApp.GetConfig("CaptureVideoCodecDisplayName", "").c_str()).Detach();
```
Comment (willkuer) [inline plugins/GSdx/GSCaptureDlg.cpp]: I don't know how bstr are handled correctly. I read a bit about it and was concerned about memory leaks. Will read more but on the first glance this seemed to be the recommended conversion from std::string to _bstr_t.

(Sorry to say but string handling is really a mess in c++. Probably it is completely clear if you understand why these hundreds of string types exist but if you are used to unicode strings as default it is kind of complex.)

File: plugins/GSdx/GSState.cpp [hunk 2]
```diff
@@ -43,14 +43,14 @@ GSState::GSState()
...
+	m_mipmap = theApp.GetConfig("mipmap", true);
 
 	s_n     = 0;
-	s_dump  = !!theApp.GetConfig("dump", 0);
-	s_save  = !!theApp.GetConfig("save", 0);
-	s_savet = !!theApp.GetConfig("savet", 0);
-	s_savez = !!theApp.GetConfig("savez", 0);
-	s_savef = !!theApp.GetConfig("savef", 0);
+	s_dump  = theApp.GetConfig("dump", false);
+	s_save = theApp.GetConfig("save", false);
```
Comment (willkuer) [inline plugins/GSdx/GSState.cpp]: oops. Needs alignment

**Comment by willkuer** [comment]
Needs to be cleaned and rebased. Part of it is in #1018. if #1018 doesnt get accepted this wont be as well. I will reopen it if reasonable.

**Pull request closed without merging**